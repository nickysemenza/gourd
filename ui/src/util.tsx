import {
  Ingredient,
  SectionIngredient,
  RecipeDetail,
} from "./api/openapi-hooks/api";
import { RecipeWrapper, TimeRange } from "./api/openapi-fetch";
import parse from "parse-duration";
import { RecipeWrapperInput } from "./api/openapi-fetch/models/RecipeWrapperInput";

export const getIngredient = (
  si: Partial<SectionIngredient>
): { name: "" } | RecipeDetail | Ingredient => {
  if (si.recipe) {
    return si.recipe;
  } else if (si.ingredient) {
    return si.ingredient.ingredient;
  }
  return { name: "" };
};

export const formatText = (text: React.ReactText) => {
  // regexr.com/5mt55
  const re = /[\d]+ ?F/g;

  if (typeof text === "number") {
    return text;
  }

  let pairs = [];
  const matches = [...text.matchAll(re)];
  if (matches.length === 0) {
    return text;
  }

  let lastProcessed = 0;
  for (const match of matches) {
    const matchStart = match.index || 0;
    const matchEnd = matchStart + match[0].length;
    pairs.push(text.substring(lastProcessed, matchStart));
    pairs.push(
      <code className="text-red-800 mx-1" key={matchStart}>
        {text.substring(matchStart, matchEnd)}
      </code>
    );
    lastProcessed = matchEnd;
  }
  pairs.push(text.substring(lastProcessed));
  return pairs;
};

const formatSeconds = (seconds: number) => {
  // https://stackoverflow.com/a/6312999
  const secs = Math.round(seconds);
  const h = Math.floor(secs / (60 * 60));
  const divisor_for_minutes = secs % (60 * 60);
  const m = Math.floor(divisor_for_minutes / 60);
  const s = Math.ceil(divisor_for_minutes % 60);

  let vals = [];
  vals.push(h > 0 ? `${h} hr` : null);
  vals.push(m > 0 ? `${m} min` : null);
  vals.push(s > 0 ? `${s} sec` : null);
  return vals.join(" ");
};
export const formatTimeRange = (range?: TimeRange) => {
  if (!range) return "";
  const { min, max } = range;
  let items = [formatSeconds(min)];
  if (max > 0 && max !== min) {
    items.push(" - ", formatSeconds(max));
  }
  return items.join("");
};

export const parseTimeRange = (input: string): TimeRange | null => {
  const parts = input.split(" - ");
  if (parts.length === 0 || parts.length > 2) return null;
  return {
    min: (parse(parts[0]) || 0) / 1000,
    max: ((parts.length === 2 && parse(parts[1])) || 0) / 1000,
  };
};

export const sumTimeRanges = (ranges: (TimeRange | undefined)[]): TimeRange => {
  let totalDuration: TimeRange = { min: 0, max: 0 };
  ranges.forEach((r) => {
    if (!!r) {
      totalDuration.min += r.min;
      totalDuration.max += r.max || r.min; // if max is 0, it means it's a finite and not a range, so use min
    }
  });

  return totalDuration;
};

export const Code: React.FC = ({ children }) => (
  <code className="rounded-sm px-2 py-0.5 bg-gray-100 text-red-500 h-6 flex text-sm font-bold w-min">
    {children}
  </code>
);

export const blankRecipeWrapperInput = (
  name: string = "",
  id: string = ""
): RecipeWrapperInput => ({
  detail: {
    name,
    quantity: 0,
    unit: "",
    sections: [],
  },
  id,
});
export const blankIngredient = (name: string): Ingredient => ({ name, id: "" });

export const scaledRound = (x: number) => x.toFixed(x < 10 ? 2 : 0);
