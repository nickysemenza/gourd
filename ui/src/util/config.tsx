import { fetchGetConfig } from "../api/react-query/gourdApiComponents";

export const getConfig = async () => {
  const config = await fetchGetConfig({});
  return config;
};
