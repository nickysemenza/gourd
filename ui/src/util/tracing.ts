import { FetchInstrumentation } from "@opentelemetry/instrumentation-fetch";
import { WebTracerProvider } from "@opentelemetry/sdk-trace-web";
import { CollectorTraceExporter } from "@opentelemetry/exporter-collector";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { ZoneContextManager } from "@opentelemetry/context-zone";
import { JaegerPropagator } from "@opentelemetry/propagator-jaeger";
import { Resource } from "@opentelemetry/resources";
import { diag, DiagConsoleLogger, DiagLogLevel } from "@opentelemetry/api";
import {
  BatchSpanProcessor,
  SimpleSpanProcessor,
} from "@opentelemetry/sdk-trace-base";

const serviceName = "gourd-ui";
const resource = new Resource({ "service.name": serviceName });
const provider = new WebTracerProvider({ resource });

export const registerTracing = (url: string, batch: boolean) => {
  if (url === "") return;
  console.info("enabled tracing", url);
  diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.ERROR);

  const exporter = new CollectorTraceExporter({
    url: url,
    // https://github.com/open-telemetry/opentelemetry-js/issues/2321#issuecomment-889861080
    headers: {
      "Content-Type": "application/json",
    },
  });

  registerInstrumentations({
    instrumentations: [
      new FetchInstrumentation({
        propagateTraceHeaderCorsUrls: /.+/,
      }),
    ],
  });
  provider.register({
    contextManager: new ZoneContextManager(),
    // propagator: new B3Propagator(),
    propagator: new JaegerPropagator(),
  });
  if (batch) {
    provider.addSpanProcessor(new BatchSpanProcessor(exporter));
  } else {
    provider.addSpanProcessor(new SimpleSpanProcessor(exporter));
  }
};

export const tracer = provider.getTracer(serviceName);
