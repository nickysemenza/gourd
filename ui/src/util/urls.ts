export const getBaseURL = () => import.meta.env.VITE_API_URL;
export const getTracingURL = () => import.meta.env.VITE_TRACING_URL || "";
export const getAPIURL = () => getBaseURL() + "/api";
