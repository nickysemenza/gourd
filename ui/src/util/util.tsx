import {
  Ingredient,
  SectionIngredient,
  IngredientKind,
} from "../api/openapi-hooks/api";
import { RecipeWrapperInput } from "../api/openapi-fetch/models/RecipeWrapperInput";

export const getIngredient = (si: Partial<SectionIngredient>) => {
  let name = "";
  let kind: IngredientKind = "ingredient";
  if (si.recipe) {
    name = si.recipe.name;
    kind = "recipe";
  } else if (si.ingredient) {
    name = si.ingredient.ingredient.name;
  }
  return { name, kind };
};

export const blankRecipeWrapperInput = (
  name = "",
  id = ""
): RecipeWrapperInput => ({
  detail: {
    name,
    quantity: 0,
    unit: "",
    sections: [],
    tags: [],
  },
  id,
});
export const blankIngredient = (name: string): Ingredient => ({ name, id: "" });

export const scaledRound = (x: number) => x.toFixed(x < 10 ? 2 : 0);
