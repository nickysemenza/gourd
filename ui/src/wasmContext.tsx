import { createContext, useEffect, useState } from "react";

export type { RichItem, Amount } from "./wasm/gourd_rs";
export type wasm = typeof import("./wasm/gourd_rs");

export const WasmContext = createContext<wasm | undefined>(undefined);

export const WasmContextProvider: React.FC<{
  children?: React.ReactNode;
}> = ({ children }) => {
  // const [cursor, setCursor] = useState({ active: false });
  const [state, setState] = useState<wasm>();
  useEffect(() => {
    const fetchWasm = async () => {
      console.time("wasm-load");
      const wasm = await import("./wasm/gourd_rs");
      setState(wasm);
      console.timeEnd("wasm-load");
    };
    fetchWasm();
  }, []);

  return <WasmContext.Provider value={state}>{children}</WasmContext.Provider>;
};
