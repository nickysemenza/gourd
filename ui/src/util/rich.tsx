import { wasm, RichItem } from "./wasmContext";

export const formatRichText = (w: wasm, text: RichItem[]) => {
  return text.map((t, x) => {
    if (t.kind === "Text") {
      return t.value;
    } else if (t.kind === "Ing") {
      return (
        <div
          className="inline text-orange-800 m-0 underline decoration-grey decoration-solid"
          key={x + "a"}
        >
          {t.value}
        </div>
      );
    } else if (t.kind === "Measure") {
      const val = t.value.pop();
      if (!val) {
        return null;
      }
      if (val.unit === "whole") {
        val.unit = "";
      }
      if (val.value === null) {
        val.value = 0;
      }
      return (
        <div
          className="inline text-green-800 m-0 underline decoration-grey decoration-solid"
          key={x}
        >
          {/* // sum_time_amounts converts the units to the best scaled */}
          {w.format_amount(val)}
        </div>
      );
    } else {
      return null;
    }
  });
};
