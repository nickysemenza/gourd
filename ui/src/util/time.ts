import { Amount, RecipeSection } from "../api/openapi-hooks/api";
import { wasm } from "./wasmContext";

export const sumTimeRanges = (
  w: wasm,
  ranges: (Amount | undefined)[]
): Amount => {
  const ranges2 = ranges.filter((r) => r !== undefined) as Amount[];
  return w.sum_time_amounts(ranges2);
};

export const getTotalDuration = (w: wasm, sections: RecipeSection[]) =>
  sumTimeRanges(
    w,
    (sections || []).map((s) => s.duration).filter((t) => t !== undefined)
  );

export const formatTimeRange = (w?: wasm, range?: Amount) => {
  // sum_time_amounts converts the units to the best scaled
  return w && range ? w.format_amount(w.sum_time_amounts([range])) : "";
};
