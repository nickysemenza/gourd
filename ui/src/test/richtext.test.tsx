import React from "react";
import { assert, expect, test } from "vitest";
import renderer from "react-test-renderer";
import { wasm } from "../util/wasmContext";
import { formatRichText } from "../util/rich";
const RichFormatter: React.FC<{
  text: string;
  ingredients: string[];
  w: wasm;
}> = ({ text, ingredients, w }) => formatRichText(w, w.rich(text, ingredients));

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();
  expect(result).toBeDefined();
  expect(result).toBeInstanceOf(Array);
  return result as renderer.ReactTestRendererJSON[];
}

const w = await import("../wasm/gourd_wasm");

test("rich text is displayed properly", () => {
  const component = renderer.create(
    <RichFormatter
      text="heat oven to 150°F for 5 minutes and put 1 cup flour in a bowl"
      w={w}
      ingredients={["flour"]}
    />
  );

  const tree = toJson(component);
  expect(tree).toMatchSnapshot();

  const children = tree.map((t) => t.children).filter((t) => t !== undefined);
  console.log(tree);
  assert.deepEqual(children, [["150 °f"], ["5 minutes"], ["1 cup"], ["flour"]]);
});
