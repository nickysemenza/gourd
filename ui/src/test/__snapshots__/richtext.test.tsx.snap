// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rich text is displayed properly 1`] = `
[
  "heat oven to ",
  <div
    className="inline text-green-800 m-0 underline decoration-grey decoration-solid"
  >
    150 Â°f
  </div>,
  " for ",
  <div
    className="inline text-green-800 m-0 underline decoration-grey decoration-solid"
  >
    5 minutes
  </div>,
  " and put ",
  <div
    className="inline text-green-800 m-0 underline decoration-grey decoration-solid"
  >
    1 cup
  </div>,
  " ",
  <div
    className="inline text-orange-800 m-0 underline decoration-grey decoration-solid"
  >
    flour
  </div>,
  " in a bowl",
]
`;
